File: java/algorithms/src/main/java/com/tatumalenko/algorithms/java/leetcode/FlattenBinaryTreeToLinkedList114.java
package com.tatumalenko.algorithms.java.leetcode;

import com.tatumalenko.algorithms.java.datastructures.BinaryTreeNode;

public class FlattenBinaryTreeToLinkedList114 {

    public TreeNode flatten(final TreeNode root) {
        return flattenIterative(root);
    }

    private TreeNode flattenIterative(final TreeNode root) {
        TreeNode cur = root;
        while (cur != null) {
            if (cur.left != null) {
                TreeNode last = (TreeNode) cur.left;
                while (last.right != null) {
                    last = (TreeNode) last.right;
                }
                last.right = cur.right;
                cur.right = cur.left;
                cur.left = null;
            }
            cur = (TreeNode) cur.right;
        }
        return root;
    }

    private TreeNode flattenRecursive(final TreeNode root) {
        if (root == null) {
            return null;
        }
        // flattenHelper(root);
        return flattenHelper(root);
    }

    private TreeNode flattenHelper(final TreeNode node) {
        if (node == null) {
            return null;
        }
        final TreeNode left = flattenHelper((TreeNode) node.left);
        final TreeNode right = flattenHelper((TreeNode) node.right);
        node.left = null;
        node.right = left;
        TreeNode temp = node;
        while (temp.right != null) {
            temp = (TreeNode) temp.right;
        }
        temp.right = right;
        return node;
    }

    static class TreeNode extends BinaryTreeNode<Integer> {

        TreeNode(final Integer val) {
            super(val);
        }

        TreeNode(final Integer val, final TreeNode left, final TreeNode right) {
            super(val, left, right);
        }

    }

}

File: java/algorithms/src/main/kotlin/com/tatumalenko/algorithms/kotlin/leetcode/FlattenBinaryTreeToLinkedList114.kt
package com.tatumalenko.algorithms.kotlin.leetcode

import com.tatumalenko.algorithms.kotlin.ds.TreeNode

object FlattenBinaryTreeToLinkedList114 {
    fun flatten(root: TreeNode?): Unit {
        flattenRightIterative(root)
    }

    private fun flattenRightIterative(tip: TreeNode?): Unit {
        var current = tip

        while (current != null) {
            current.left?.let { left ->
                val rightest = findLowestRightNode(left)
                rightest.right = current?.right
                current?.right = current?.left
                current?.left = null
            }
            current = current.right
        }
        
    }

    private fun flattenLeftRecursive(tip: TreeNode?): Unit {
        tip?.let { node ->
            node.left?.let { flattenLeftRecursive(it) }

            node.right?.let { flattenLeftRecursive(it) }

            node.right?.let {
                val leftest = findLowestLeftNode(node)
                leftest.left = node.right
                node.right = null
            }
        }
    }

    private fun flattenRightRecursive(tip: TreeNode?): Unit {
        tip?.let { node ->
            node.left?.let { flattenRightRecursive(it) }

            node.right?.let { flattenRightRecursive(it) }

            node.left?.let {
                val newNodeRight = it
                val rightest = findLowestRightNode(newNodeRight)
                rightest.right = node.right
                node.right = newNodeRight
                node.left = null
            }
        }
    }

    private fun findLowestLeftNode(node: TreeNode): TreeNode {
        var current: TreeNode? = node
        while (current?.left != null) {
            current = current.left
        }
        return current ?: node
    }

    private fun findLowestRightNode(node: TreeNode): TreeNode {
        var current: TreeNode? = node
        while (current?.right != null) {
            current = current.right
        }
        return current ?: node
    }

    fun <E> MutableList<E>.pop(): E? {
        return if (this.count() > 0) this.removeAt(this.count() - 1) else null
    }
}